### PR TITLE
Assertions in hardware

### PR DESCRIPTION
**Pull request by tilk**
This PR adds a function named `assertion` which allows to define hardware-level ("run-time") assertions. Failed assertion will fail a test.

TODO:
* [x] Write a test
* [x] Add docstrings
* [x] Handle assertions in regression tests too (pysim engine)
* [x] Assertions with source locations (to help easily find the failed assertion)
* [x] Make independent from Coreblocks (requires #544)

### CODE REVIEW

**Comment by lekcyjna123** [comment]
Maybe move the `gp` to `gen_params` and other naming changes to separate PR?

**Comment by tilk** [comment]
> Maybe move the `gp` to `gen_params` and other naming changes to separate PR?

These are crucial for this to work, but you're right, this PR would be easier to review this way.

File: test/common/coreblocks.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+__all__ = ["CoreblocksTestCaseWithSimulator", "make_assert_handler"]
+
+
+def make_assert_handler(gen_params: GenParams, my_assert: Callable[[int, str], Any]):
+    def assert_handler():
+        yield Passive()
+        while True:
+            if (yield assert_bit(gen_params)):
+                for v, (n, i) in assert_bits(gen_params):
+                    my_assert((yield v), f"Assertion at {n}:{i}")
```
Comment (lekcyjna123) [inline test/common/coreblocks.py:20]: Is it possible, that the assertion will be False? If any of `v` is False, then `assert_bit` is False, so we don't take the `if` and we will skip the condition checking.

**Review by lekcyjna123 — COMMENTED** [review]
Now I see that we should probably move `DependencyManager` to transactron, as assertions aren't something coreblocks specific and they need the manager.

File: test/common/coreblocks.py (hunk 1 above)
Comment (tilk) [reply test/common/coreblocks.py:20]: There is a negation missing in the condition. (Why didn't you say so?)

**Comment by tilk** [comment]
> Now I see that we should probably move `DependencyManager` to transactron, as assertions aren't something coreblocks specific and they need the manager.

No, they're not, but they should be syntactically simple to use, and currently the easiest way to do so is to use GenParams.

File: test/common/coreblocks.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+
+__all__ = ["CoreblocksTestCaseWithSimulator", "make_assert_handler"]
+
+
+def make_assert_handler(gen_params: GenParams, my_assert: Callable[[int, str], Any]):
+    def assert_handler():
+        yield Passive()
+        while True:
+            yield Delay(5e-7)  # Shorter than clock cycle
```
Comment (lekcyjna123) [inline test/common/coreblocks.py:17]: Take length of clock cycle as an argument and calculate `Delay` by removing epsilon from it.

How this handler should be ordered regarding to TransactionProfiles? I think that order of this two things shouldn't matter. I am right?

File: test/common/coreblocks.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            if not (yield assert_bit(gen_params)):
+                for v, (n, i) in assert_bits(gen_params):
+                    my_assert((yield v), f"Assertion at {n}:{i}")
+            yield
+
+    return assert_handler
+
+
+class CoreblocksTestCaseWithSimulator(TestCaseWithSimulator):
```
Comment (lekcyjna123) [inline test/common/coreblocks.py:26]: After moving DependencyManager to transactron it is not needed to have this handler in coreblocks specific file. We can made it part of `TestCaseWithSimulator`.

File: test/utils/test_assertion.py
Comment (lekcyjna123) [inline test/utils/test_assertion.py:1]: I would propose to put it into `test/transactron/utils/test_assertion.py`

File: test/common/coreblocks.py (hunk 3 above)
Comment (tilk) [reply test/common/coreblocks.py:26]: But then the field containing the `DependencyManager` should no longer be named `gen_params`, which will provoke an awkward change in all the Coreblocks tests.

**Review by lekcyjna123 — APPROVED** [review]
Overall it looks great. Only two small comments left.

File: transactron/utils/assertion.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+__all__ = ["AssertKey", "assertion", "assert_bit", "assert_bits"]
+
+
+@dataclass(frozen=True)
+class AssertKey(ListKey[tuple[Value, SrcLoc]]):
+    pass
+
+
+def assertion(value: Value, *, src_loc_at: int = 0):
+    """Signal if assertion false.
```
Comment (piotro888) [inline transactron/utils/assertion.py:19]: This title may be misleading - if signal/condition is true then assertion is held, not false.

File: transactron/utils/assertion.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    Parameters
+    ----------
+    dependencies : DependentCache | DependencyManager
+        The assertion feature uses the `DependencyManager` to store
+        assertions. If `DependentCache` is passed instead, `DependencyManager`
+        is requested from the cache.
+    """
+    if isinstance(dependencies, DependentCache):
+        dependencies = dependencies.get(DependencyManager)
```
Comment (piotro888) [inline transactron/utils/assertion.py:53]: Is passing `DependentCache` useful? `assertion` always adds to `DependencyContext`

**Review by piotro888 — APPROVED** [review]
LGTM

File: transactron/utils/assertion.py (hunk 5 above)
Comment (tilk) [reply transactron/utils/assertion.py]: It was a leftover from previous iterations.

File: transactron/utils/assertion.py (hunk 4 above)
Comment (tilk) [reply transactron/utils/assertion.py]: Rewrote the docstring a bit.

**Pull request merged**